projects/distributed-tracing/src/shared/graphql/model/schema/filter/global-graphql-filter.service.ts

```diff
@@ -0,0 +1,27 @@
+import { Injectable } from '@angular/core';
+import { GraphQlFilter } from './graphql-filter';
+
+@Injectable({ providedIn: 'root' })
+export class GlobalGraphQlFilterService {
+  private readonly filterRules: Map<symbol, FilterRule> = new Map();
+
+  public getGlobalFilters(scope: string): GraphQlFilter[] {
+    return Array.from(this.filterRules.values()).flatMap(rule => rule.filtersForScope(scope));
+  }
+
+  public mergeGlobalFilters(scope: string, localFilters: GraphQlFilter[] = []): GraphQlFilter[] {
```

Comment:
Can this be used with table - filter bar interaction?

Reply:
Not sure what you mean - will it work with with the filter bar (yes), or could we use the filter bar to do this instead of using the handlers (no, because this applies in places without a filter bar).

Reply:
So we have `FilterUrlService` that keeps a filter bar and url in sync. I think table depends on it either directly or indirectly (via urls) to construct and apply filters in table requests. With the new service, I was thinking what if we combine the two so that only one service is responsible for applying filters in graphql.

Reply:
So I think there's potential for a wider scale refactor, but there's nothing today that would solve the centralization issue. Filter bar and filter url service basically work, as you said, to keep that one component in sync with the url. The trickier part is to how to get those filters into queries. Right now, the filter bar is wired into certain dashboards by applying those filters (from the url or user interaction) directly to the dashboard's root data source. This approach really only works on homogenous screens (we really only use the filter bar on single-widget dashboards like service list), because it depends on every data source being able to apply the same filter. Even then, each data source has to opt in to merging the inherited filters - which is even more error prone, as data sources are much more numerous and change more frequently.

Given the time to do it, I think we could unify the systems, but it would be a major undertaking. I'm thinking of a hierarchical system of query contexts (not necessarily limited to filters), with a root that is applying global things like this, and with new contexts created underneath by things like filter bars - but even in that system, somewhere we need a way of applying that context in a way that is specific to each request, as they all have different shapes - and that, I think belongs on the handler side, not the requesting side.

Reply:
Sounds like a good project for you :)

Reply:
hah - if only I had the time